### DISCUSSION TITLE
Will Air publish to Html5?

### DISCUSSION THREAD

**Discussion by fftidus** [airsdk/Adobe-Runtime-Support#642] [Ideas]
Apple has banned the html5 technology of APP,so other h5 engines can't publish app for Apple store next year!
If air can publish html5,it can be the  true cross-platform engine.

**Comment by neuronix** [comment]
Please read the FAQ: https://airsdk.harman.com/faq

To keep the issues list clean, please close your issue and next time check https://forum.starling-framework.org/ first where you will generally find quick & good answers :)

**Comment by rewb0rn** [comment]
As the year closes, it would be a great time to get an update on this or an update on the future of the Air roadmap in general. Html5 target would definitely be on our wishlist, apart from keeping the current platform targets alive and well (Harman is already doing a great job with this).

Meanwhile, porting our codebase from AS3 to Haxe has proven to be a good choice to both provide swf / Air targets as well as HTML5.

**Comment by ajwfrost** [comment]
Briefly..

Support for the other platforms is coming! slower than I had hoped, we've hit a few snags but are getting through them with support from Adobe. And we're looking forward to some of the changes we can then bring to the platform to address some of the bugs and add a few tweaks and new features in the APIs.

In terms of HTML based support, we have an interesting discussion going on with a third party that might be able to help enable this. We should know more around March/April time.

thanks

**Reply by Fancy2209** [reply]
Have there been any updates about this?

**Reply by ajwfrost** [reply]
Slowly .. we've had a few more discussions, had a bit of a play with a few ideas, and have some more conversations already lined up. The WebAssembly world has moved on a bit which has helped, so we may be able to work on some basic support for specific types of content i.e. focusing on Stage3D, and maybe some 'gpu' mode rendering using WebGL. Still fairly early stages but it's becoming more likely I think.

**Reply by raresn** [reply]
@ajwfrost can you imagine being able to export to browser again? Holy cow. this would be HUGE.

**Reply by marchbold** [reply]
@raresn It would be massive boost to the platform :)

**Reply by Ender22** [reply]
I'd do a backflip in happiness, and that's not even physically possible, but I'd do it!

**Reply by Fancy2209** [reply]
While this isn't ready, ruffle works really well nowdays, mind you though it doesn't support AIR specific functionality as it's a FP Reimplantation, not AIR

**Comment by tuarua** [comment]
> Support for the other platforms is coming!

@ajwfrost Is there an ETA on AIR being updated so that it is compiled against iOS SDK 13 ?
I have hit a blocker because it is still 12 and despite setting platformsdk I am getting linking issues against a Swift static library.

**Comment by matrix3d** [comment]
?2022

**Reply by ajwfrost** [reply]
From a technical perspective, things have moved forwards a bit, but there are some other hurdles that we will need to work through. It's still on our minds and we're trying to work out a practical way that we can provide this....

**Comment by Shaunmax** [comment]
Here is a tutorial on how to convert your Flash/AIR project to Haxe : - 
https://www.youtube.com/watch?v=JTJqxXVB6rU&t=340s

**Comment by Elintondm** [comment]
Excellent tutorial! Too bad it only works for those who use starling. Anyone knows how to migrate to html5 from a project without starling?

**Reply by Shaunmax** [reply]
Just ignore the starling part, the remaining should be same!

**Comment by Matvii-Jarosh** [comment]
How are things going in 2024? I dream of seeing Adobe Air in the browser

**Reply by Ender22** [reply]
This would be #1 for me as well. Would solve issues that keep me up at night like lacking Huawei support for their future non-android OS's. Also allow for more integration of my app into various web-based portals.

**Reply by raresn** [reply]
Browser support would be a huge thing. Maybe Harman has something in the pipeline. Do you have any updates in this direction there @ajwfrost ?

**Reply by ajwfrost** [reply]
It's something we definitely want to do! We have had discussions with a company that's a world leader in WebAssembly and believe we know how we'd do it; it's likely that we'd need to do this as an AOT-based solution (like iOS) because of some licensing issues but that should also help with performance. 

It's a big-ticket item though, and there are a few others of those that we're looking at currently (i.e. AOT on Android, and support for Nintendo Switch). I think we'd be able to look into this after the AOT on Android work but as ever, there continue to be a significant number of unplanned changes that we have to make because of the continuous improvements that Apple and Google make to their platforms - which takes our bandwidth away somewhat. It's definitely not forgotten though...!

In terms of the approach, we'd not be able to support all of the normal APIs, due to running in the browser sandbox. The plan would be to focus on Stage3D support perhaps before we look at some of the internal/software rendering mechanisms, and to implement the APIs gradually based on requests/feedback. In discussions with Adobe about it, they'd felt the memory management would be a key challenge, but taking support from our WebAssembly experts it seems like their solutions/compilers would overcome that.

So - watch this space, I wish I could be more specific in timings, but I'll give updates when I can!

thanks

**Reply by ghost** [reply]
AOT on Android is huge too✨

**Reply by Fancy2209** [reply]
> It's something we definitely want to do! We have had discussions with a company that's a world leader in WebAssembly and believe we know how we'd do it; it's likely that we'd need to do this as an AOT-based solution (like iOS) because of some licensing issues but that should also help with performance.
> 
> It's a big-ticket item though, and there are a few others of those that we're looking at currently (i.e. AOT on Android, and support for Nintendo Switch). I think we'd be able to look into this after the AOT on Android work but as ever, there continue to be a significant number of unplanned changes that we have to make because of the continuous improvements that Apple and Google make to their platforms - which takes our bandwidth away somewhat. It's definitely not forgotten though...!
> 
> In terms of the approach, we'd not be able to support all of the normal APIs, due to running in the browser sandbox. The plan would be to focus on Stage3D support perhaps before we look at some of the internal/software rendering mechanisms, and to implement the APIs gradually based on requests/feedback. In discussions with Adobe about it, they'd felt the memory management would be a key challenge, but taking support from our WebAssembly experts it seems like their solutions/compilers would overcome that.
> 
> So - watch this space, I wish I could be more specific in timings, but I'll give updates when I can!
> 
> thanks

Wouldn't at least most APIs from the flash package work? Flash emulators like Ruffle and AwayFL have been allowing them to be used inside Browser Sandboxes for a long time now, granted, I have no idea how the code side compares to the real AIR

**Reply by neuronix** [reply]
> i.e. AOT on Android, and support for Nintendo Switch

@ajwfrost Is there much of a demand for Switch support ? It wouldn't hurt but I tend to think web would be much more of a priority for most developers.

In my opinion, web support is probably the only feature that could keep AIR alive long term. 

I doubt many developers will continue investing into the AIR platform instead of using other more popular solutions like Unity, Defold, Godot for games... or React, Expo etc. for apps...  when Flash & AIR's biggest historical advantage over the competition - true crossplatform support (web, desktop & mobile) - is gone.

**Reply by Fancy2209** [reply]
> > i.e. AOT on Android, and support for Nintendo Switch
> 
> @ajwfrost Is there much of a demand for Switch support ? It wouldn't hurt but I tend to think web would be much more of a priority for most developers.
> 
> In my opinion, web support is probably the only feature that could keep AIR alive long term.
> 
> I doubt many developers will continue investing into the AIR platform instead of using other more popular solutions like Unity, Defold, Godot for games... or React, Expo etc. for apps... when Flash & AIR's biggest historical advantage over the competition - true crossplatform support (web, desktop & mobile) - is gone.

I feel like Switch and Consoles is a great idea, I'm gonna be honest, I've tried other engines and Flash just feels nicer for me for some reason
Also Switch allows devs to put their old AS3 games on consoles

**Reply by ajwfrost** [reply]
@neuronix I agree with you re the longer term need for this, absolutely. There have been (and still are) some political considerations though..

My understanding is that games on consoles have much better revenue figures so for some of the premium games that use AIR, it would be quite compelling to be able to run these on a console. For the "old AS3 games" thing - whilst individuals who have small/basic games are unlikely to get very far (I suspect) in persuading Nintendo to accept these, it's more likely that an aggregator would be able to set up this sort of thing..

**Reply by raresn** [reply]
To be honest, I thought the same about starting with the browser first. However, from a financial perspective, it makes sense. Still, I'll definitely be popping open a bottle of champagne when AIR is back in the browser!

Thank you for the insight, @ajwfrost. If you have any updates on the developments, please know that we eagerly consume every bit of information you share with us.

Although it's not the main topic of this thread, do you have any news on the TypeScript idea instead of AS3? Are there any roadmaps or plans for AIR in general that you can share?

Thanks a lot!

**Reply by shaucky** [reply]
Maybe starting a new thread for news about platform/language support is better? Or community can only pick up information by irregular asking?
 
Currently we know Harman is working on compile to WebAssembly, Nintendo Switch, but actually i think everything is going too slowly… This thread is almost five years now.
 
Now we can build wonderful apps/games on traditional platforms, but since AIR released on 2008, number of platform supported had no change. From my perspective, the most intuitive comparison is the difference between the platforms supported by AIR and Unity.
 
I just want to say that support for new platforms is always anticipated, but when will we see the products? (not mean to blame, but worry)

**Reply by GeneralVimes** [reply]
Here is a python script what I wrote for myself: https://github.com/GeneralVimes/AS32JS/
It converts my games well (AIR Starling to HTML5 Phaser)

**Comment by marcussacana** [comment]
2025 and we keep going... 6 years already

**Reply by GeneralVimes** [reply]
Try my Python scripts - will they work for you? https://github.com/generalVimes/AS32JS

**Reply by marcussacana** [reply]
Unless it has a reimplementation of the flash API it will not work for me, the project is a bit big.

**Reply by Fancy2209** [reply]
> Unless it has a reimplementation of the flash API it will not work for me, the project is a bit big.

You can use OpenFL and Apache Royale if you don't depend on swf assets
But yeah it's unfortunate how hard it's been to HARMAN to get WASM out. It's understandable but I wish it hadn't been taking so long

**Reply by marcussacana** [reply]
Yeah, it has a lot of swf assets :/ 
No impossible convert the images in swfs, but the sprites and animations not sure of what can be done.

**Comment by Matvii-Jarosh** [comment]
2026?

**Reply by neuronix** [reply]
> @neuronix I agree with you re the longer term need for this, absolutely. There have been (and still are) **some political considerations** though..

**Reply by Matvii-Jarosh** [reply]
Happy 7th birthday, Adobe Air in the browser

**Reply by ajwfrost** [reply]
Yes... we had some interesting discussions last year about the approach for this, the challenge is to get the right people available to work on this without having distractions from other projects (which tends to happen, particularly over the last 8-10 months with some other Adobe activities..)

We have a few 'big ticket' items that we want to get done, that people are looking for. AIR in a browser is one of the bigger ones, there's a possible shortcut if we used a virtualisation technology such as CheerpX although it would need some updates for Stage3D support. So whilst I can't promise it, I will try to have someone at least get a basic proof of concept running so we can see what sort of performance we'd get out of that kind of mechanism...

thanks

**Comment by megajogos** [comment]
You can use Ruffle to run the app directly in the browser.
Our app runs very well with Ruffle: https://www.megajogos.com.br/app/ruffle_app.jsp
Ruffle — https://ruffle.rs/
I would prefer Harman to focus on reducing ANRs on Android rather than adding browser support, since Ruffle already solves that use case.
